### ISSUE TITLE
Configuration for compiled files is not found

### ISSUE THREAD

**Issue by dan1994** [microsoft/vscode-makefile-tools#173]
I keep getting notified that configuration for files can't be found. This happens every time I open a source/header file, and is accompanied by an intrusive appearance of the output window. It may also result in incorrect intellisense, but I'm not sure about that.

This is the warning I get:
![image](https://user-images.githubusercontent.com/6225230/118267706-29562300-b4c5-11eb-8067-2a85748fb863.png)

I use a hierarchy of makefiles to compile different parts of the project (bootloader and kernel), and a single makefile that invokes `make -C` on their respective directories:

```makefile
.PHONY: $(BOOTLOADER)
$(BOOTLOADER):
	make -C $(SRC_DIR)/boot

.PHONY: $(KERNEL)
$(KERNEL):
	make -C $(SRC_DIR)/kernel
```

In the internal makefiles I use pattern rules so that filenames are not explicitly specified:

```makefile
$(BUILD_DIR)/%.c.o: %.c
	mkdir -p $(@D)
	$(GCC) $(CFLAGS) -c $< -o $@
```

System Information:
Windows 10 Build 19042.964
WSL 2

VSCode:
Version: 1.56.2 (user setup)
Commit: 054a9295330880ed74ceaedda236253b4f39a335
Date: 2021-05-12T17:13:13.157Z
Electron: 12.0.4
Chrome: 89.0.4389.114
Node.js: 14.16.0
V8: 8.9.255.24-electron.0
OS: Windows_NT x64 10.0.19042

MakefileTools version 0.2.2

**Comment by GiancarloFusiello** [comment]
I'm also seeing this appear every time I open a `.cpp` file in the editor.

For context, I'm using the Remote Containers extension and open the project inside of a Ubuntu 20.04 container.

Any guidance for suppressing the output window to appear would be helpful. 

Thanks.

**Comment by andreeis** [comment]
@dan1994, even if the file names are not mentioned in the makefile, the extension needs to see them in the dry-run log. Can you share the output of the command "make --dry-run --always-make --keep-going --print-directory"? If the files are not explicitly logged in there, the extension cannot work properly. Are you using any silencing directives?


@dan1994 and @GiancarloFusiello , are the files that cause this message part of the project you loaded in VSCode? If you open a random source file from somewhere else, which is not built by your currently opened project, they won't be listed in the dry-run log and it's by design. If you have a source file that is part of your project and you get that message, then it's a bug either in our extension or in how you wrote your makefiles or maybe you chose a limited target (instead of "all" or one more inclusive of all your files) in the Makefile Tools left side panel.

Also, for the case of opening a source file from outside, which by design we don't expect Makefile Tools to provide IntelliSense configuration for, what would you prefer to happen with the message? We thought that bringing up the "Output Channel" window will make the users notify that there is a configuration problem and they would investigate it (fix it in their makefiles or send us a GitHub report). Otherwise, without helping the users notice such problem, they may operate on wrong or missing IntelliSense for a longer time than necessary.

We can add a setting that would always prevent the output channel from becoming visible. But do you really need that? I think it's risky, such setting can easily be forgotten to a value that would prevent the output channel from becoming visible when issues happen.
Maybe we can implement something per source file, like an "ignore not found configuration for file" state that you would set file by file as the extension discovers and complains about them?

**Comment by andreeis** [comment]
@dan1994 and @GiancarloFusiello, also is this something that regressed recently or the message "CppTools will set a default configuration" was always present for these files, if you ever used Makefile Tools older than 0.2.2?

**Comment by dan1994** [comment]
I ran the command and it doesn't explicitly mention the files, probably because of the use of `make -C`. Here is the output:

```
make: Entering directory '/home/dan/projects/l/kernel_from_scratch'
make -C src/boot
make -C src/kernel
rm -f bin/os.bin
dd if=bin/boot/boot.bin >> bin/os.bin
dd if=bin/kernel/kernel.bin >> bin/os.bin
dd if=/dev/zero bs=512 count=100 >> bin/os.bin
make: Leaving directory '/home/dan/projects/l/kernel_from_scratch'
```

> Are you using any silencing directives?

No.

> are the files that cause this message part of the project you loaded in VSCode?

Yes.

> maybe you chose a limited target

This is the correct target.

> Also, for the case of opening a source file from outside, which by design we don't expect Makefile Tools to provide IntelliSense configuration for, what would you prefer to happen with the message?

A good question. On the one hand I would like to know if a specific file is not covered by intellisense, but on the other hand popping the output window is really annoying, and I would rather reserve it for fatal makefile errors. Can this be shown as an error so it will appear on the problems tab, and show a squiggly line on the first line? 

Anyways, if you do stick with the format of a message I would suggest emphasizing that this is an error and supply possible causes (file not in vscode project), as the current message is not indicative enough of a problem.

> is this something that regressed recently or the message "CppTools will set a default configuration" was always present for these files, if you ever used Makefile Tools older than 0.2.2?

I haven't used this extension recently so I don't have any inputs regarding this.

**Comment by dan1994** [comment]
I decided to check if `make` can do a dry run recursively and ran into [this](https://stackoverflow.com/questions/8554101/make-recursive-dry-run) post that advises to replace invocations of `make` with `$(MAKE)`, and it solves the problem. 

You might want to suggest that to users, as I think this might be a common use case.

I do get the error message once for the first header file I open (per reload of vscode), which seems like a different bug.

**Comment by andreeis** [comment]
@dan1994, thank you for your ideas, very useful points. We will address each of them.
About getting the error for the first header that is opened in VSCode, that is now by design. Since headers are in general included by C/C++ sources instead of compiled alone, their IntelliSense exists only in the context of the translation unit that references them. Notice that as soon as you open in VSCode a source file that includes that header, the squiggles in the header will be resolved.

**Comment by GiancarloFusiello** [comment]
@dan1994 I've tried replacing `make` for `$(MAKE)` but that does not seem to make a difference in my case.

@andreeis I'm working on a very simple project as a method to get to grips with VSCode and C++ so I'm happy to share. [Here](https://github.com/GiancarloFusiello/CppND-System-Monitor) is the repo I'm working on.

This is what the Makefile looks like at present:
```makefile
.PHONY: all
all: format build

.PHONY: format
format:
	clang-format src/* include/* -i

.PHONY: build
build:
	mkdir -p build
	cd build && \
	cmake .. && \
	$(MAKE)

.PHONY: debug
debug:
	mkdir -p build
	cd build && \
	cmake -DCMAKE_BUILD_TYPE=debug .. && \
	$(MAKE)

.PHONY: clean
clean:
	rm -rf build

.PHONY: launch
launch:
	docker-compose run --rm --service-ports system-monitor
	docker-compose down --volumes --remove-orphans

```

dryrun.log
```
make --dry-run --keep-going --print-directory
make: Entering directory '/workspaces/CppND-System-Monitor'
mkdir -p build
cd build && \
cmake .. && \
make
 
-- Configuring done
 
-- Generating done
 
-- Build files have been written to: /workspaces/CppND-System-Monitor/build
 
make[1]: Entering directory '/workspaces/CppND-System-Monitor/build'
/usr/bin/cmake -S/workspaces/CppND-System-Monitor -B/workspaces/CppND-System-Monitor/build --check-build-system CMakeFiles/Makefile.cmake 0
 
/usr/bin/cmake -E cmake_progress_start /workspaces/CppND-System-Monitor/build/CMakeFiles /workspaces/CppND-System-Monitor/build/CMakeFiles/progress.marks
make -f CMakeFiles/Makefile2 all
 
make[2]: Entering directory '/workspaces/CppND-System-Monitor/build'
make -f CMakeFiles/monitor.dir/build.make CMakeFiles/monitor.dir/depend
 
make[3]: Entering directory '/workspaces/CppND-System-Monitor/build'
cd /workspaces/CppND-System-Monitor/build && /usr/bin/cmake -E cmake_depends "Unix Makefiles" /workspaces/CppND-System-Monitor /workspaces/CppND-System-Monitor /workspaces/CppND-System-Monitor/build /workspaces/CppND-System-Monitor/build /workspaces/CppND-System-Monitor/build/CMakeFiles/monitor.dir/DependInfo.cmake --color=
make[3]: Leaving directory '/workspaces/CppND-System-Monitor/build'
 
make -f CMakeFiles/monitor.dir/build.make CMakeFiles/monitor.dir/build
 
make[3]: Entering directory '/workspaces/CppND-System-Monitor/build'
/usr/bin/cmake -E cmake_echo_color --switch= --green --progress-dir=/workspaces/CppND-System-Monitor/build/CMakeFiles --progress-num=2 "Building CXX object CMakeFiles/monitor.dir/src/linux_parser.cpp.o"
/bin/x86_64-linux-gnu-g++-9   -I/workspaces/CppND-System-Monitor/include  -g   -Wall -Wextra -std=gnu++17 -o CMakeFiles/monitor.dir/src/linux_parser.cpp.o -c /workspaces/CppND-System-Monitor/src/linux_parser.cpp
 
/usr/bin/cmake -E cmake_echo_color --switch= --green --bold --progress-dir=/workspaces/CppND-System-Monitor/build/CMakeFiles --progress-num=8 "Linking CXX executable monitor"
/usr/bin/cmake -E cmake_link_script CMakeFiles/monitor.dir/link.txt --verbose=
make[3]: Leaving directory '/workspaces/CppND-System-Monitor/build'
/usr/bin/cmake -E cmake_echo_color --switch= --progress-dir=/workspaces/CppND-System-Monitor/build/CMakeFiles --progress-num=1,2,3,4,5,6,7,8 "Built target monitor"
make[2]: Leaving directory '/workspaces/CppND-System-Monitor/build'
/usr/bin/cmake -E cmake_progress_start /workspaces/CppND-System-Monitor/build/CMakeFiles 0
make[1]: Leaving directory '/workspaces/CppND-System-Monitor/build'
make: Leaving directory '/workspaces/CppND-System-Monitor'
 
```

targets.log
```
make all --print-data-base --no-builtin-variables --no-builtin-rules --question
# GNU Make 4.2.1
# Built for x86_64-pc-linux-gnu
# Copyright (C) 1988-2016 Free Software Foundation, Inc.
# License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
# This is free software: you are free to change and redistribute it.
# There is NO WARRANTY, to the extent permitted by law.

# Make data base, printed on Mon May 17 08:40:05 2021

# Variables

# automatic
<D = $(patsubst %/,%,$(dir $<))
# automatic
?F = $(notdir $?)
# default
.SHELLFLAGS := -c
# environment
VSCODE_AGENT_FOLDER = /root/.vscode-server
# environment
VSCODE_LOG_NATIVE = false
# automatic
?D = $(patsubst %/,%,$(dir $?))
# automatic
@D = $(patsubst %/,%,$(dir $@))
# automatic
@F = $(notdir $@)
# makefile
CURDIR := /workspaces/CppND-System-Monitor
# makefile
SHELL = /bin/sh
# environment
VSCODE_NLS_CONFIG = {"locale":"en","availableLanguages":{}}
# makefile (from 'Makefile', line 1)
MAKEFILE_LIST :=  Makefile
# 'override' directive
GNUMAKEFLAGS := 
# environment
VSCODE_VERBOSE_LOGGING = true
# environment
VSCODE_CWD = /vscode/vscode-server/bin/x64/054a9295330880ed74ceaedda236253b4f39a335
# default
MAKE_HOST := x86_64-pc-linux-gnu
# environment
PATH = /vscode/vscode-server/bin/x64/054a9295330880ed74ceaedda236253b4f39a335/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
# environment
VSCODE_LOG_STACK = false
# environment
VSCODE_IPC_HOOK_CLI = /tmp/vscode-ipc-2074041a-b00a-4e81-8d58-d852117926dc.sock
# default
.FEATURES := target-specific order-only second-expansion else-if shortest-stem undefine oneshell archives jobserver output-sync check-symlink load
# environment
SSH_AUTH_SOCK = /tmp/vscode-ssh-auth-ca8629e6197f6553ec54e362a34327898989370b.sock
# automatic
%F = $(notdir $%)
# environment
VSCODE_PIPE_LOGGING = true
# environment
PWD = /workspaces/CppND-System-Monitor
# default
.LOADED := 
# environment
VSCODE_AMD_ENTRYPOINT = vs/server/remoteExtensionHostProcess
# environment
HOME = /root
# environment
REMOTE_CONTAINERS = true
# environment
APPLICATION_INSIGHTS_NO_DIAGNOSTIC_CHANNEL = true
# automatic
+F = $(notdir $+)
# environment
VSCODE_HANDLES_UNCAUGHT_ERRORS = true
# automatic
^D = $(patsubst %/,%,$(dir $^))
# environment
HOSTNAME = e0532e95e4a7
# environment
MAKELEVEL := 0
# default
MAKE = $(MAKE_COMMAND)
# default
MAKECMDGOALS := all
# default
MAKE_VERSION := 4.2.1
# makefile
.DEFAULT_GOAL := all
# automatic
%D = $(patsubst %/,%,$(dir $%))
# environment
REMOTE_CONTAINERS_IPC = /tmp/vscode-remote-containers-ipc-ca8629e6197f6553ec54e362a34327898989370b.sock
# default
MAKE_COMMAND := make
# default
.VARIABLES := 
# automatic
*F = $(notdir $*)
# makefile
MAKEFLAGS = pqrR
# environment
MFLAGS = -pqrR
# automatic
*D = $(patsubst %/,%,$(dir $*))
# automatic
+D = $(patsubst %/,%,$(dir $+))
# environment
BROWSER = /vscode/vscode-server/bin/x64/054a9295330880ed74ceaedda236253b4f39a335/bin/helpers/browser.sh
# default
MAKEFILES := 
# automatic
<F = $(notdir $<)
# environment
LC_ALL = C
# environment
DEBIAN_FRONTEND = noninteractive
# automatic
^F = $(notdir $^)
# default
SUFFIXES := 
# environment
REMOTE_CONTAINERS_SOCKETS = ["/tmp/vscode-ssh-auth-ca8629e6197f6553ec54e362a34327898989370b.sock"]
# default
.INCLUDE_DIRS = /usr/include /usr/local/include /usr/include
# environment
VSCODE_INJECT_NODE_MODULE_LOOKUP_PATH = /vscode/vscode-server/bin/x64/054a9295330880ed74ceaedda236253b4f39a335/remote/node_modules
# default
.RECIPEPREFIX := 
# environment
LANG = C
# environment
VSCODE_EXTHOST_WILL_SEND_SOCKET = true
# variable set hash-table stats:
# Load=60/1024=6%, Rehash=0, Collisions=5/90=6%

# Pattern-specific Variable Values

# No pattern-specific variable values.

# Directories

# . (device 121, inode 56851525): 19 files, no impossibilities.

# 19 files, no impossibilities in 1 directories.

# Implicit Rules

# No implicit rules.

# Files

all: build
#  Phony target (prerequisite of .PHONY).
#  Command line target.
#  Implicit rule search has not been done.
#  File does not exist.
#  File has been updated.
#  Needs to be updated (-q is set).
# variable set hash-table stats:
# Load=0/32=0%, Rehash=0, Collisions=0/5=0%

.PHONY: all format build debug clean launch
#  Implicit rule search has not been done.
#  Modification time never checked.
#  File has not been updated.

# Not a target:
.SUFFIXES:
#  Implicit rule search has not been done.
#  Modification time never checked.
#  File has not been updated.

format:
#  Phony target (prerequisite of .PHONY).
#  Implicit rule search has not been done.
#  File does not exist.
#  File has not been updated.
#  recipe to execute (from 'Makefile', line 6):
	clang-format src/* include/* -i

# Not a target:
Makefile:
#  Implicit rule search has been done.
#  Last modified 2021-05-17 08:39:54.227842975
#  File has been updated.
#  Successfully updated.

debug:
#  Phony target (prerequisite of .PHONY).
#  Implicit rule search has not been done.
#  File does not exist.
#  File has not been updated.
#  recipe to execute (from 'Makefile', line 17):
	mkdir -p build
	cd build && \
	cmake -DCMAKE_BUILD_TYPE=debug .. && \
	$(MAKE)

build:
#  Phony target (prerequisite of .PHONY).
#  Implicit rule search has not been done.
#  Implicit/static pattern stem: ''
#  File does not exist.
#  File has been updated.
#  Needs to be updated (-q is set).
# automatic
# @ := build
# automatic
# % := 
# automatic
# * := 
# automatic
# + := 
# automatic
# | := 
# automatic
# < := 
# automatic
# ^ := 
# automatic
# ? := 
# variable set hash-table stats:
# Load=8/32=25%, Rehash=0, Collisions=2/13=15%
#  recipe to execute (from 'Makefile', line 10):
	mkdir -p build
	cd build && \
	cmake .. && \
	$(MAKE)

# Not a target:
.DEFAULT:
#  Implicit rule search has not been done.
#  Modification time never checked.
#  File has not been updated.

clean:
#  Phony target (prerequisite of .PHONY).
#  Implicit rule search has not been done.
#  File does not exist.
#  File has not been updated.
#  recipe to execute (from 'Makefile', line 24):
	rm -rf build

launch:
#  Phony target (prerequisite of .PHONY).
#  Implicit rule search has not been done.
#  File does not exist.
#  File has not been updated.
#  recipe to execute (from 'Makefile', line 28):
	docker-compose run --rm --service-ports system-monitor
	docker-compose down --volumes --remove-orphans

# files hash-table stats:
# Load=10/1024=1%, Rehash=0, Collisions=0/39=0%
# VPATH Search Paths

# No 'vpath' search paths.

# No general ('VPATH' variable) search path.

# strcache buffers: 1 (0) / strings = 29 / storage = 259 B / avg = 8 B
# current buf: size = 8162 B / used = 259 B / count = 29 / avg = 8 B

# strcache performance: lookups = 60 / hit rate = 51%
# hash-table stats:
# Load=29/8192=0%, Rehash=0, Collisions=1/60=2%
# Finished Make data base on Mon May 17 08:40:05 2021
```

**Comment by GiancarloFusiello** [comment]
This is what the output window is displaying when I open any `.cpp` file:
```
Generating configuration cache with command: make --dry-run --keep-going --print-directory
Writing the dry-run output: /workspaces/CppND-System-Monitor/.vscode/dryrun.log
Generating dry-run elapsed time: 0.694
Preprocessing: "/workspaces/CppND-System-Monitor/.vscode/dryrun.log"
Preprocess elapsed time: 0
Parsing for IntelliSense.
Updating the CppTools IntelliSense Configuration Provider.
Parsing dry-run output for CppTools Custom Configuration Provider.
Sending configuration for file /workspaces/CppND-System-Monitor/src/linux_parser.cpp -----------------------------------
Parsing for IntelliSense elapsed time: 0.046
Parsing for launch targets.
Found the following 1 new launch targets defined in the makefile: /workspaces/CppND-System-Monitor>a.out()
Complete list of launch targets: /workspaces/CppND-System-Monitor>a.out()
Parsing for launch targets elapsed time: 0.01
Generating parse content for build targets.
Generating targets information with command: make all --print-data-base --no-builtin-variables --no-builtin-rules --question
Writing the dry-run output: /workspaces/CppND-System-Monitor/.vscode/targets.log
Generating dry-run elapsed time: 0.043
Parsing for build targets from: "/workspaces/CppND-System-Monitor/.vscode/targets.log"
Found the following 6 new build targets defined in the makefile: all;build;clean;debug;format;launch
Complete list of build targets: all;build;clean;debug;format;launch
Parsing build targets elapsed time: 0.006
Configure finished. The status for all the subphases that ran:
generateParseContent: return code = 0, elapsed time = 0.694
preprocessParseContent: return code = 0, elapsed time = 0
parseIntelliSense: return code = 0, elapsed time = 0.046
parseLaunch: return code = 0, elapsed time = 0.01
dryrunTargets: return code = 1, elapsed time = 0.043
parseTargets: return code = 0, elapsed time = 0.006
Configure succeeded.
Configure elapsed time: 0.813
Configuration for file /workspaces/CppND-System-Monitor/src/system.cpp was not found. CppTools will set a default configuration.
Configuration for file /workspaces/CppND-System-Monitor/src/format.cpp was not found. CppTools will set a default configuration.
Configuration for file /workspaces/CppND-System-Monitor/src/main.cpp was not found. CppTools will set a default configuration.

```

**Comment by andreeis** [comment]
We are glad that everyone is unblocked.
I will keep this work item open for these comments I made above.

> We can add a setting that would always prevent the output channel from becoming visible. But do you really need that? I think it's risky, such setting can easily be forgotten to a value that would prevent the output channel from becoming visible when issues happen.
> Maybe we can implement something per source file, like an "ignore not found configuration for file" state that you would set file by file as the extension discovers and complains about them?

**Comment by hitman-codehq** [comment]
@andreeis I have also been having problems with IntelliSense not working on my project.  Luckily I stumbled upon this issue here and this helped me to figure out that it is my makefile that was the problem.  I have this rule:

```
$(OBJ)/%.o: %.c
	@echo Compiling $<...
	$(CC) $(CFLAGS) $(IFLAGS) -o $(OBJ)/$*.o $<

```

It turns out to be those ... characters in the echo line that cause the problem as they prevent the .c file from being found.  The output in dryrun.log is:

echo Compiling MyFile.c...

When I add a space before the ... then it works:

echo Compiling MyFile.c ...

I didn't see any mention in the documentation (https://marketplace.visualstudio.com/items?itemName=ms-vscode.makefile-tools) of the filenames needing to be printed somewhere in the dryrun.log file (without trailing ...!) in order for IntelliSense to work.  IntelliSense configuration with the makefile plugin is rather black magic at the moment, so could I suggest a bit more explanation of how it works in the documentation?  It would help people to debug their problems.

Even better, when outputting the dry run information in the output window, if no files are found, output a message at the end telling the user this, and giving them tip that the names of the files should be listed in the dryrun.log output.  I see that there is positive confirmation when a file is found (Sending configuration for file \cygdrive\d\Source\ClassAct\Examples\Label\LabelExample.c -----------------------------------) but it would be good to also have negative confirmation.

Handling those trailing ... would be nice too, if possible.  That had me scratching my head for a while.  :-)

**Comment by gcampbell-msft** [comment]
Reading through this issue, it seems like we have fixed or provided workarounds for the issues mentioned, with that in mind, I'm going to go ahead and close this issue. 

However, @Florin-Popescu, @dan1994, @hitman-codehq, please feel free to open a new issue if I'm incorrectly closing this and there are still issues or feature requests that you have. Thank you very much!
